Remove more IsDisabled checks

reaperrr opened the pull request:
Depends on #13998.

This removes most of the remaining `Actor.IsDisabled` look-ups.
After this, only `Gate`,  `ProductionQueue` and AI `AirStates` remain (might be worth investigating whether those are easy enough to migrate to add them to this PR, too).

reaperrr committed:
Remove Actor.IsDisabled check from Cloak

Makes cloaking only disableable via conditions.

reaperrr committed:
Remove IsDisabled from AffectsShroud traits

Makes them only disableable via conditions.

reaperrr committed:
Adapt TS to removal of actor IsDisabled check in Cloak

reaperrr committed:
Replace turret IsDisabled check with AT.IsTraitDisabled check

reaperrr committed:
Make Attack* traits PausableConditional and replace IsDisabled with IsTraitDisabled checks

reaperrr committed:
Adapt TS to removal of IsDisabled checks from Attack* traits

GraionDilach commented:
Wouldn't https://github.com/OpenRA/OpenRA/pull/14202/commits/c7b7d8f874e3eb2c6cf69ae051fb3408296a53cf conflict with the ammo refactors?

pchote commented:
> OpenRA.Mods.Common/Traits/AffectsShroud.cs(70,19): error CS0103: The name `IsDisabled' does not exist in the current context
> 
> OpenRA.Mods.Common/Traits/AffectsShroud.cs(73,87): error CS0103: The name `cachedDisabled' does not exist in the current context
> 
> OpenRA.Mods.Common/Traits/AffectsShroud.cs(105,38): error CS0103: The name `cachedDisabled' does not exist in the current context
> 
> OpenRA.Mods.Common/Traits/Turreted.cs(99,13): error CS0103: The name `self' does not exist in the current context
> 
> Compilation failed: 4 error(s), 0 warnings

pchote commented on OpenRA.Mods.Common/Traits/Turreted.cs:
```diff
@@ -16,7 +16,7 @@
 namespace OpenRA.Mods.Common.Traits
 {
-	public class TurretedInfo : ITraitInfo, UsesInit<TurretFacingInit>, Requires<BodyOrientationInfo>, IActorPreviewInitInfo
+	public class TurretedInfo : ITraitInfo, UsesInit<TurretFacingInit>, Requires<BodyOrientationInfo>, Requires<AttackTurretedInfo>, IActorPreviewInitInfo
```
I really don't like the idea of making this depend on `AttackTurreted`.  Can we do this a different way?

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackBase.cs:
```diff
@@ -81,7 +81,7 @@ protected virtual Func<IEnumerable<Armament>> InitializeGetArmaments(Actor self)
 		protected virtual bool CanAttack(Actor self, Target target)
 		{
-			if (!self.IsInWorld || IsTraitDisabled || self.IsDisabled())
+			if (!self.IsInWorld || IsTraitDisabled)
```
Shouldn't this return false on `IsTraitPaused` too?

reaperrr commented:
Updated and rebased.

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackFollow.cs:
```diff
@@ -84,7 +84,7 @@ public override Activity Tick(Actor self)
 					return NextActivity;
 
-				if (self.IsDisabled())
+				if (attack.IsTraitDisabled)
```
AttackBase maps `IsDisabled()` &rarr; `IsTraitPaused`, so shouldn't that be the same here?

pchote commented on OpenRA.Mods.Common/Traits/Attack/AttackTurreted.cs:
```diff
@@ -17,17 +17,22 @@ namespace OpenRA.Mods.Common.Traits
 
 	public class AttackTurreted : AttackFollow
 	{
+		public readonly AttackTurretedInfo AttackTurretedInfo;
```
This doesn't look very nice.  Just cast it when you need it instead?

reaperrr replied:
Not sure what you mean with casting?
I couldn't come up with anything else that compiled and allowed looking up this AttackTurreted-specific Info in Turreted.

pchote commented on OpenRA.Mods.Common/Traits/Turreted.cs:
```diff
@@ -113,38 +114,30 @@ void ITick.Tick(Actor self)
-			}
-			else
-			{
-				realignTick = 0;
```
What happened to this line?

reaperrr commented:
Updated.

pchote commented:
Regression: The RA gap generator does not disable on low power.

pchote commented:
Regression: RA tesla coils (and other powered defenses) lose the ability to target and their range circles when powered down.  Something seems to be treating them as disabled rather than paused.

reaperrr committed:
Fix regressions introduced by #13998

reaperrr commented:
> Regression: The RA gap generator does not disable on low power.
> Regression: RA tesla coils (and other powered defenses) lose the ability to target and their range circles when powered down. Something still seems to be treating them as disabled rather than paused.

Fixed. These were introduced by #13998, but at least some of them needed Attack* traits to be pausable to fix them, so this is the right place to fix them anyway.

pchote commented:
I'm not sure if this is a new regression or and old bug, but if you build an AA gun (or another turreted defense), order it to force attack a Mig (or another unit) and then immediately disable it then the turret will continue to track while disabled, without firing.

pchote commented:
Something else that i'm not sure about: in some cases disabling an Attack* seems to cancel the target, and in others not.

GraionDilach commented:
The turrets-stuck-in-attack mode bug (which the last two instances have showcased) have been already filed as #12856.

reaperrr commented:
> Something else that i'm not sure about: in some cases disabling an Attack* seems to cancel the target, and in others not.

I assume *disabling* should always cancel it?

pchote commented:
Sorry, I meant disabling in the sense of EMP or power-down ingame. i.e. Pausing.

reaperrr commented:
> Edit: This is indeed a regression vs the last release.
Edit: this is also a regression vs the last release.

This doesn't make any sense to me. Going by code changes, I see no way how `PauseOnCondition` should behave any different to Actor.IsDisabled on last release.

Both 'regressions' should already be present on last release, going by the code changes in this PR alone.

reaperrr committed:
Fix turrets following target even if AttackTurreted is disabled or paused

reaperrr commented:
Ok, managed to find the cause of the first one. Still investigating the 2nd one.

reaperrr commented:
Updated.

pchote commented on mods/ts/rules/gdi-vehicles.yaml:
```diff
@@ -364,7 +369,7 @@ JUGG:
 		Voice: Attack
 		Armaments: deployed
-		RequiresCondition: deployed
+		PauseOnCondition: empdisable || !deployed
```
Shouldn't this be
```
RequiresCondition: deployed
PauseOnCondition: empdisable
```
?

pchote commented on mods/ts/rules/nod-vehicles.yaml:
```diff
@@ -138,7 +140,7 @@ TTNK:
 	AttackFrontal:
 		Voice: Attack
-		RequiresCondition: undeployed
+		RequiresCondition: !empdisable && undeployed
```
Shouldn't these be 
```
RequiresCondition: undeployed
PauseOnCondition: empdisable
```
and
```
RequiresCondition: deployed
PauseOnCondition: empdisable
```
?  We do want to switch these off completely when in the opposite deploy state.

pchote commented on mods/ts/rules/nod-vehicles.yaml:
```diff
@@ -251,7 +253,7 @@ ART2:
 		Voice: Attack
 		Armaments: deployed
-		RequiresCondition: deployed
+		PauseOnCondition: empdisable || undeployed
```
Likewise here.

pchote reviewed (commented):
The latest version seems to have fixed all the regressions in the classic mods, great job!

I found a couple of possible issues in TS:
* Deployed Tick Tanks and Juggernauts can't attack.
* Inline comments below

reaperrr commented:
Updated.

> Deployed Tick Tanks and Juggernauts can't attack.

Found and fixed the problem. `AttackTurreted`'s new `Turrets` property defaults to "primary", while the turrets of those 2 + arty had their name changed to `deployed`.

reaperrr commented:
@abcdefg30 @penev92 There's one last follow-up waiting that will remove the remaining 3 uses of `Actor.IsDisabled()`, so this is also one of the PRs that should get merged soon so we can finally wrap this up.

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/babcab0fd80d2d73a4f3565ed0caad6fe7cd55ef...3441324c2160290ca56f10257a8b3f1827e80012
